Return empty result for packages that are not available on the feed.

erdembayar opened the pull request:
## Bug
Signed packages return null stream, instead of try to process just return default empty value.
Fixes: https://github.com/NuGet/Home/issues/9728
Regression: Yes/No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: Details_about_the_fix  

## Testing/Validation

Tests Added: Yes/No  
Reason for not adding tests:  
Validation:

erdembayar committed:
Return empty result for signed package.

nkolev92 commented on src/NuGet.Core/NuGet.Protocol/Resources/PackageMetadataResourceV3.cs:
```diff
@@ -149,6 +149,11 @@ private async Task<T> DeserializeStreamDataAsync<T>(Stream stream, CancellationT
 
+            if (stream == null)
+            {
+                return default(T);
```
Test please? :)

erdembayar replied:
Added 2 unit test which correspond with below 2 cases in HttpSource.cs.

![image](https://user-images.githubusercontent.com/8766776/86267447-9564bc80-bb7b-11ea-967e-20a44aa299e9.png)

zivkan commented on src/NuGet.Core/NuGet.Protocol/Resources/PackageMetadataResourceV3.cs:
```diff
@@ -149,6 +149,11 @@ private async Task<T> DeserializeStreamDataAsync<T>(Stream stream, CancellationT
         {
             token.ThrowIfCancellationRequested();
 
+            if (stream == null)
```
Did you determine why the stream is null?  How did the old code work?

erdembayar replied:
It allows null stream then just try next source stream until it hit non-null return value or exhaust of all sources.
![image](https://user-images.githubusercontent.com/8766776/86267692-f4c2cc80-bb7b-11ea-9938-6d2dcfd305f2.png)

Here are actual streams behind each source. For signed package case it goes into last source which is  handled in PackageExtractor.cs

{https://api.nuget.org/v3/index.json}
{https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json}
{https://pkgs.dev.azure.com/azure-public/vside/_packaging/vs-impl/nuget/v3/index.json}
{https://dotnet.myget.org/F/roslyn/api/v3/index.json}
{https://dotnet.myget.org/F/nuget-volatile/api/v3/index.json}
{https://pkgs.dev.azure.com/azure-public/vside/_packaging/vssdk/nuget/v3/index.json}
{https://dotnet.myget.org/F/msbuild/api/v3/index.json}
{https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json}
{
![image](https://user-images.githubusercontent.com/8766776/86267960-53884600-bb7c-11ea-9c20-6c7179985bff.png)
}

zivkan replied:
None of those URLs should be returning empty streams.  In fact, opening them in my browser they all return valid json data.

erdembayar replied:
But they don't have our local signed package then simply returns from HttpSource.cs line 143. 
Actual stream of course exist but they don't have our "signed package" then it returns null stream to signal that with HttpSourceResultStatus.NotFound. Here null stream implicit way of signaling your package not found in this source please try next one.

![image](https://user-images.githubusercontent.com/8766776/86270965-f6db5a00-bb80-11ea-8025-9d7df7f127a5.png)

nkolev92 replied:
Ok, sounds like this is a simple case of "the package is not on the feed". 
I imagine this happens quite often in the PM UI. 

Is the exception swallowed in the Package manager UI?

erdembayar replied:
There is no exception, only implicit null passed around to signal package is not here until it hit source with meta data.
At last source "Signed package " it finds meta 

![image](https://user-images.githubusercontent.com/8766776/86280674-18dcd880-bb91-11ea-8545-650d033748b6.png)

Signed package found here once LocalFolderUtility.GetPackageV2 scans "SignedPackage" local path.
![image](https://user-images.githubusercontent.com/8766776/86280830-69543600-bb91-11ea-9412-0369817b4078.png)

erdembayar replied:
One reason I didn't think through this problem was none of my projects I worked had locally created nuget source.

zivkan replied:
so `stream == null` only for local sources, not http sources? Your previous comments didn't make that clear.  However in this case your tests don't make a lot of sense to me, as they're returning results with `HttpStatus.NotFound`, so it feels like it's not testing the same thing.

In any case I'm ok with the change because it's reproducing what the old code did:
https://github.com/NuGet/NuGet.Client/blob/5d1af18e560b5f381626b5a7e9ce2acacb6211db/src/NuGet.Core/NuGet.Protocol/Utility/StreamExtensions.cs#L21-L26

I'm not sure if the miscommunication in this thread was due to lack of understanding, or just poor communication, but I think we should try to improve our communication. Also, I would prefer if you could provide links to the code on github, rather than showing screenshots of code, because it allows me to get more context and look at code that's not shown in your screenshot. Several of your screenshots don't even tell me which file the code is from.

erdembayar replied:
@zivkan Sorry for miscommunication due to lack of my understanding for local repository. Myself never encountered with local nuget source before. Yes, it's using null stream as implicit way of telling package not found for httpsources, so eventually it fall back to local source.  I'll use github link next time, still trying to adjust to github environment.

nkolev92 replied:
I'm still not sure what the exact root cause is here. 
Would the same thing happen if we had 2 different http sources, one contains the package we are looking for and the other one not containing it?

erdembayar replied:
One without package will just return null and other will return actual data.  By chance I didn't hit this error, I should have discovered it before. It looks here "local" sources is not really important, just if there are multiple source or not and order of them.  Why I didn't discover this was I always working with nuget.org(which usually first http source in order) which has almost everything or solution with 1 source.

nkolev92 replied:
Makes sense. Approved based on this comment.

erdembayar committed:
Add Unit test for handling Null stream if package not found.

erdembayar committed:
Keep only 1 concise unit test.

erdembayar committed:
Add unit test to handle null stream if no content from http response.

erdembayar committed:
Revert unneeded change.

donnie-msft reviewed (approved):
Your tests mention NoContent, but I thought the problem was about a Signed package with an Untrusted Signer. 
So, does that mean Untrusted Signer signed packages cause NoContent, and that's why your tests make sense? Is that the main scenario where we could have a NoContent package?
If yes&yes, I'm good with the changes.
If no, let's discuss.

erdembayar commented:
> Your tests mention NoContent, but I thought the problem was about a Signed package with an Untrusted Signer.
> So, does that mean Untrusted Signer signed packages cause NoContent, and that's why your tests make sense? Is that the main scenario where we could have a NoContent package?
> If yes&yes, I'm good with the changes.
> If no, let's discuss.

I added "NoContent" one to cover case in HttpSource.cs line 146. I didn't hit that condition so far (I guess it happens due to bad package or some http connection problem), just added it for completeness of covering all cases. For signed package it return from line 143.

![image](https://user-images.githubusercontent.com/8766776/86269620-ddd1a980-bb7e-11ea-8288-c4629c9b666d.png)

erdembayar committed:
Rename after Donnie comment.

nkolev92 reviewed (approved):
Make sure you post your answer to my question. 

LGTM beyond that :)
